### ISSUE TITLE
support typing for non-direct subclasses of charmbase in `@trace_charm`

### ISSUE THREAD

**Issue by PietroPasotti** [canonical/tempo-k8s-operator#145]
Another argument for the approach is that, the `trace_charm` decorator has type annotation for `CharmBase` ([ref](https://github.com/canonical/tempo-k8s-operator/blob/fab9528391d4c2934dfb6359a06e28a0a4c01ece/lib/charms/tempo_k8s/v1/charm_tracing.py#L460))

I'm all for type annotations, but for charms that implement [some other inheritance](https://github.com/canonical/mysql-operator/blob/858e07fa655c8ed0f8428fdff129cd50b86bc0aa/src/charm.py#L139), this breaks LSP(pyright) goodies

_Originally posted by @paulomach in https://github.com/canonical/tempo-k8s-operator/issues/128#issuecomment-2183619281_

**Comment by PietroPasotti** [comment]
@paulomach can you show what typing error(s) you're seeing by the way?
pyright seems to be fine with subclasses

![image](https://github.com/canonical/tempo-k8s-operator/assets/6230162/cca2cc39-ba01-4dfc-936a-09b87ed0ce30)

Also decorating the superclass instead seems to be good.

**Comment by paulomach** [comment]
Hi @PietroPasotti , the following:

```python
from typing import Type

class A:
    def foo(self):
        print("original foo")

def decorator():
    # def _decorator(cls):  <- untyped arg
    def _decorator(cls: Type[A]):
        setattr(cls, "_decorated", True)
        return cls

    return _decorator

@decorator()
class B(A):
    def foo(self):
        print("foo")

    def bar(self):
        print("bar")


if __name__ == "__main__":
    b = B()
    if b._decorated:
        print("b is decorated")
    b.foo()
    b.bar()  # here pyright tells "bar" is unknown. Removing Type on _decorator allows correct resolution
```

**Comment by paulomach** [comment]
@PietroPasotti this is fixed by [PR#144](https://github.com/canonical/tempo-k8s-operator/pull/144) already, can close.
